packages/preview/paddling-tongji-thesis/0.1.1/paddling-tongji-thesis/elements.typ

```diff
@@ -0,0 +1,168 @@
+  place("|", dx: -1.6cm, dy: 21.5cm)
+}
+
+#let empty-par() = {
+  v(-1.2em)
+  box()
+}
+
+#let make-cover(cover) = align(center)[
+  #image("figures/tongji.svg", height: 2.25cm)
+  #text(
+    "TONGJI UNIVERSITY",
+    font: font-family.hei,
+    size: font-size.at("-2"),
+    weight: "bold",
+  )
+
+  #v(30pt)
+  #text("本科毕业设计（论文）", font: font-family.hei, size: font-size.at("-0"))
+  #v(60pt)
+
+  #set text(font: font-family.hei, size: font-size.at("-2"))
+  #grid(
+    columns: (5em, auto),
+    gutter: 16pt,
+    ..cover.enumerate().map(((idx, value)) => {
+      set text(size: font-size.at("-2"))
+      if calc.even(idx) {
+        let arr = value.clusters()
+        let k = (4 - arr.len()) / (arr.len() - 1)
+        arr.join([#h(1em * k)])
+      } else {
+        block(
+          width: 100%,
+          inset: 4pt,
+          stroke: (bottom: 1pt + black),
+          align(center, value),
+        )
+      }
+    }),
+  )
+]
+
+#let make-abstract(title: "", abstract: "", keywords: (), prompt: (), is-english: false) = {
+  align(
+    center,
+  )[
+    #v(1em)
+    #text(font: font-family.hei, size: font-size.at("-2"), weight: "bold", title)
+    #v(1.5em)
+  ]
+  heading(prompt.at(0), numbering: none, outlined: false)
+
+  set par(first-line-indent: 2em)
+  text(abstract)
+
+  v(5mm)
+  set par(first-line-indent: 0em)
+  text(font: font-family.xiaobiaosong, weight: "bold", prompt.at(1))
+  let keywords-string = if is-english {
+    keywords.join(", ")
+  } else {
+    keywords.join("，")
+  }
+  text(keywords-string)
+}
+
+#let make-outline(title: "目录", depth: 3, indent: true) = {
+  // outline(title: "目录", depth: 3)
+  heading(title, numbering: none, outlined: false)
+  set par(first-line-indent: 0pt, leading: 0.9em)
+  locate(
+    it => {
+      let elements = query(heading.where(outlined: true), it)
+
+      for el in elements {
+        // Skip headings that are too deep
+        if depth != none and el.level > depth { continue }
+
+        let el_number = if el.numbering != none {
+          numbering(el.numbering, ..counter(heading).at(el.location()))
+          h(0.5em)
+        }
+
+        let line = {
+          if indent {
+            let indent-width = if el.level == 1 {
+              0pt
+            } else if el.level == 2 {
+              1em
+            } else if el.level == 3 {
+              4em
+            } else {
+              0pt
+            }
+
+            h(indent-width)
+          }
+
+          link(el.location(), el_number)
+          link(el.location(), el.body)
+          box(width: 1fr, h(0.25em) + box(width: 1fr, repeat[·#h(1pt)]) + h(0.25em))
+          link(el.location(),str(counter(page).at(el.location()).first()))
+
+          linebreak()
+        }
+
+        // link(el.location(), line)
+        line
+      }
+    },
+  )
+}
+
+#let heavyrulewidth = .08em
+#let lightrulewidth = .05em
+#let cmidrulewidth = .03em
+
+#let toprule(stroke: heavyrulewidth) = {
+  hlinex(stroke: stroke)
```

Comment:
`hlinex` is only imported in `tongjithesis.typ`, and is not defined in this file. You should import it again at the top of `elements.typ` too.